csrc/backend_ops/ncnn/onnx2ncnn/utils.h

```diff
@@ -0,0 +1,394 @@
+// Copyright (c) OpenMMLab. All rights reserved.
+#pragma once
+
+#include <float.h>
+#include <google/protobuf/io/coded_stream.h>
+#include <google/protobuf/io/zero_copy_stream_impl.h>
+#include <google/protobuf/message.h>
+#include <google/protobuf/text_format.h>
+#include <limits.h>
+
+#include <fstream>
+#include <iostream>
+
+#include "onnx.pb.h"
+
+/**
+ * @brief find graph node by output name
+ *
+ * @param graph
+ * @param name
+ * @return onnx::NodeProto*
+ */
+static onnx::NodeProto* find_node_by_output_name(onnx::GraphProto* mutable_graph,
+                                                 const std::string& name) {
+  const int input_count = mutable_graph->node_size();
+  for (int i = 0; i < input_count; ++i) {
+    onnx::NodeProto* node = mutable_graph->mutable_node(i);
+
+    for (int j = 0; j < node->output_size(); ++j) {
+      auto output = node->output(j);
+      if (output == name) {
+        return node;
+      }
+    }
+  }
+
+  return nullptr;
+}
+
+static bool read_proto_from_binary(const char* filepath, onnx::ModelProto* message) {
+  std::ifstream fs(filepath, std::ifstream::in | std::ifstream::binary);
+  if (!fs.is_open()) {
+    fprintf(stderr, "open failed %s\n", filepath);
+    return false;
+  }
+
+  google::protobuf::io::IstreamInputStream input(&fs);
+  google::protobuf::io::CodedInputStream codedstr(&input);
+
+#if GOOGLE_PROTOBUF_VERSION >= 3011000
+  codedstr.SetTotalBytesLimit(INT_MAX);
+#else
+  codedstr.SetTotalBytesLimit(INT_MAX, INT_MAX / 2);
+#endif
+
+  bool success = message->ParseFromCodedStream(&codedstr);
+
+  fs.close();
+
+  return success;
+}
+
+static std::vector<int> get_node_attr_ai(const onnx::NodeProto& node, const char* key) {
+  std::vector<int> v;
+
+  for (int i = 0; i < node.attribute_size(); i++) {
+    const onnx::AttributeProto& attr = node.attribute(i);
+    if (attr.name() == key) {
+      v.resize(attr.ints_size());
+      for (int j = 0; j < attr.ints_size(); j++) {
+        v[j] = std::max(std::min(attr.ints(j), (::google::protobuf::int64)INT_MAX),
+                        (::google::protobuf::int64)INT_MIN);
+      }
+
+      break;
+    }
+  }
+
+  return v;
+}
+
+static void set_node_attr_ai(onnx::NodeProto& node, const char* key,
+                             const std::vector<int>& value) {
+  onnx::AttributeProto* attr_group = node.add_attribute();
+  attr_group->set_name(key);
```

Comment:
May check if the attribute of the key is already exist.

Reply:
And, as this is `onnx2ncnn`, why do we need to set the attribute of an ONNX node?

Reply:
`fuse_conv_reshape` needs `set_node_attr_ai` to **rewrite** shape para, donot need to check the key.